### PR TITLE
Feat/multichain friendly cellars

### PR DESCRIPTION
**Pull request by crispymangoes**
This PR changes the cellar codebase so that deployments on different chains can be accomplished by only changing addresses in deploy scripts, as opposed to the current codebase which would require changing addresses in contracts.

More specifically, any hardcoded addresses were changed to be immutable constructor arguments, so that deployments on other chains can be done by creating new deployment scripts with the appropriate addresses.

In addition to the above this PR adds three new contracts, `CustomFastGasFeed.sol`, `ProtocolFeeCollector.sol`, and `CellarInitializableV2_3`.

`CustomFastGasFeed.sol` : Mimics the Chainlink Mainnet Fast Gas Feed https://data.chain.link/ethereum/mainnet/gas/fast-gas-gwei 
For the vast majority of use cases, this contract will be deployed with the owner set as the zero address. This fixes the fast gas price at 0 gwei. For the majority of L2s this is fine, because the gas is so cheap.

`ProtocolFeeCollector.sol` : Will hold any earned protocol fees in it until a more advanced fee collector is developed. This is needed because we do not have a Gravity Bridge deployment on any networks other than ETH Mainnet.

`CellarInitializableV2_3.sol` : Is a copy of `CellarInitializableV2_2.sol` except instead of initializing the strategist payout address, it initializes the Aave V3 pool address. This way Cellar's can be deployed on L2s without needing to change `CellarInitializableV2_2.sol`.

### CODE REVIEW

File: src/modules/adaptors/Morpho/MorphoAaveV2ATokenAdaptor.sol [hunk 1]
```diff
@@ -36,36 +36,45 @@ contract MorphoAaveV2ATokenAdaptor is BaseAdaptor, MorphoRewardHandler {
...
+
+    constructor(
+        address _morpho,
+        address _morphoLens,
+        uint256 minHealthFactor,
+        address rewardDistributor
+    ) MorphoRewardHandler(rewardDistributor) {
+        _verifyConstructorMinimumHealthFactor(minHealthFactor);
+        morpho = IMorphoV2(_morpho);
+        morphoLens = IMorphoLensV2(_morphoLens);
```
Comment (0xEinCodes) [inline src/modules/adaptors/Morpho/MorphoAaveV2ATokenAdaptor.sol:65]: I guess for protocols that do more chain-specific customization, interfaces may not be chain-agnostic. Just something to note.

**Review by 0xEinCodes — COMMENTED** [review]
Still going through the tests and scripts. All looks good so far! Made some comments to better understand how the design will be set up, in particular the fee collection, for multichain deployment.

File: src/modules/adaptors/Aave/AaveATokenAdaptor.sol [hunk 2]
```diff
@@ -35,41 +35,44 @@ contract AaveATokenAdaptor is BaseAdaptor {
...
-     * Normally the identifier would just be the address of this contract, but this
-     * Identifier is needed during Cellar Delegate Call Operations, so getting the address
-     * of the adaptor is more difficult.
+     * @notice The Aave V3 Pool contract on current network.
+     * @dev For mainnet use 0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2.
      */
-    function identifier() public pure override returns (bytes32) {
-        return keccak256(abi.encode("Aave aToken Adaptor V 1.2"));
-    }
+    IPool public immutable pool;
```
[the remark this answers is not in the document]
Comment (crispymangoes) [reply src/modules/adaptors/Aave/AaveATokenAdaptor.sol:46]: Correct, byte code wise, the pure functions would have the code jump to this function, grab the return value, then carry on. But using the immutable values, they are now copied into the bytecode, so there is nothing to grab, it is already there

File: src/modules/adaptors/Morpho/MorphoAaveV2ATokenAdaptor.sol (hunk 1 above)
Comment (crispymangoes) [reply src/modules/adaptors/Morpho/MorphoAaveV2ATokenAdaptor.sol:65]: Correct if a protocol has changed their interfaces, then we will have to make contract changes to support it no matter what we do

File: src/modules/ProtocolFeeCollector.sol [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+contract ProtocolFeeCollector is Owned {
+    using SafeTransferLib for ERC20;
+
+    constructor(address _owner) Owned(_owner) {}
+
+    /**
+     * @notice Implements the Gravity Bridge `sendToCosmos` function, but
+     *         just holds ERC20s in contract.
+     */
+    function sendToCosmos(address asset, bytes32, uint256 amount) external {
```
[the remark this answers is not in the document]
Comment (crispymangoes) [reply src/modules/ProtocolFeeCollector.sol:21]: So in the future, we will create a real FeeCollector contract with more intricate logic. Maybe it takes fees and just sends them to cosmos through Axelar, or it could swap all the fees into WETH, and bridge it to Ethereum, then send the fees to cosmos through the gravity bridge. This is still up in the air which is why we have this temporary solution.

TLDR in the future we will have a contract that will replace this simple FeeCollector and it will actually handle the fees when `sendToCosmos` is called.

File: src/base/CellarInitializableV2_3.sol [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    function initialize(bytes calldata params) external initializer {
+        (
+            address _owner,
+            Registry _registry,
+            ERC20 _asset,
+            string memory _name,
+            string memory _symbol,
+            uint32 _holdingPosition,
+            bytes memory _holdingPositionConfig,
+            uint64 _strategistPlatformCut,
```
[the remark this answers is not in the document]
Comment (crispymangoes) [reply src/base/CellarInitializableV2_3.sol:46]: So that amount is the percent of fees that go to the strategist. So if it is set to 0.8e18, then 80% of the fees go to the strategist, and 20% goes to the protocol. This was added as an input bc new cellars usually have different fee splits. so its a value we almost always change after deployment. But if we set it on initialization, then we dont need to worry about setting it later. Then I also added the aave pool address input because while all cellars deployed on ETH would use the same aave pool address, once we start deploying on mulitchain this is no longer the case. So we set it here on initialization. 

If you try to keep the strategist payout address as an input, then you run into stack too deep errors, and of the 3 things(The `_strategistPlatformCut`, `aavePool`, and `strategistPayoutAddress`), the `strategistPayoutAddress` is the easiest to change later because the strategist has the power to set it them selves, where as the `_strategistPlatformCut` arg must be changed by governance, and `aavePool` can ONLY be set during initialization, there is no function to change it.

**Review by 0xEinCodes — APPROVED** [review]
Finished looking through the rest. I made one comment I think, otherwise looks good!

File: src/base/CellarInitializableV2_3.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/base/CellarInitializableV2_3.sol:46]: kk that all makes sense, thanks for the clarification!
